### PR TITLE
realistic Rayleigh/Mie atmospheric scattering

### PR DESCRIPTION
**Pull request by Mc-Pain**
Experimental shaders for atmospheres on terrestrial planets.

TO-DO list:

- [x] Partial depth calculation: shadow cylinder cuts the far side of ray
- [x] Eclipse handling: use atmosDiffuse when sampling the ray instead of multiplying the overall result
- [x] Use atmospheric density in calculations
- [x] Use star luminosity. We don't want to make atmosphere bright from M-class star thousands AU away as if it was 0.5 AU away
- [ ] Add RGB matrix (For non-Nitrogen atmospheres?)

### CODE REVIEW

**Comment by nozmajner** [comment]
Could you post some screenshots, I'm curious.

**Comment by Mc-Pain** [comment]
Found the old one but shaders has been reworked since.
изображение

I'll post actual screenshots later (this evening?)

**Comment by impaktor** [comment]
Just checking: our eclipse at [Epsilon Eridani](https://wiki.pioneerspacesim.net/wiki/Places_of_interest) still works?
Also, something @Web-eWorks to check on your system-editor branch, if planets and planes and unspecified parameters are being changed.

**Comment by nozmajner** [comment]
How cool is that! Never thought it is missing until I saw what's missing.
If it is a performace hog, maybe it could be useful to tie it to make it togglable in settings?

**Comment by Mc-Pain** [comment]
Origin of this code:
https://www.scratchapixel.com/lessons/procedural-generation-virtual-worlds/simulating-sky/simulating-colors-of-the-sky.html

**Comment by Mc-Pain** [comment]
изображение
Epsilon Eridani, eclipse

**Comment by Web-eWorks** [comment]
Thanks for taking a crack at this! There are a few "problems" which are why I've not already tried to implement this before:
- Obviously, as you've mentioned, performance. Trying to brute-force Rayleigh-Mie is pretty much impossible to do outside of an offline raytracer, which is what that article was originally written for. There are plenty of papers available discussing ways to make it more performant for real-time rendering, but more on that later.
- The reference Rayleigh/Mie coefficients used by most implementations (and this one specifically) are (AFAIK) only appropriate for an oxygenated / earth-like atmosphere and will not produce visually correct results in CO2 / Methane atmospheres or in different atmospheric densities. Without some approximation to generate semi-physically-based Rayleigh / Mie coefficients there's not much point to introducing "realistic" atmospheric scattering.
- Have you verified if "aerial perspective" AKA views from space work correctly? Depending on the implementation of the Nishita model those are sometimes not handled correctly.

If you're interested in addressing the first problem while still implementing the Nishita model, you'll be interested in [this GpuGems2 chapter](https://developer.nvidia.com/gpugems/gpugems2/part-ii-shading-lighting-and-shadows/chapter-16-accurate-atmospheric-scattering) which manages to cram the necessary math down into something a little more performant.

I cannot promise this PR will be merged however, as even if performance improves on desktop our min-spec target is semi-recent laptop/integrated GPUs, and this PR would both need to address the second point above and provide this as an "experimental" toggleable option.

The other reason I can't promise to merge the PR is that we instead intend to implement `A Scalable and Production Ready Sky and Atmosphere Rendering Technique` (by Sebastien Hillaire, described [here](https://sebh.github.io/publications/egsr2020.pdf) and [here](https://blog.selfshadow.com/publications/s2020-shading-course/hillaire/s2020_pbs_hillaire_slides.pdf)) which is a physically-based model designed for real-time game rendering on everything from mobile to high-end PC. Hillaire's model provides better performance and integrates better with other atmospheric effects like cloud rendering. The implementation is not possible until required plumbing work is done in both the render API and scene rendering code sections however, so the usual caveat of "no ETA" applies.

If you are interested in getting some extremely useful code merged, starting with a way to procedurally generate at least semi-accurate Rayleigh/Mie coefficients for our various atmospheric exoplanets would be very worthwhile regardless of which rendering model gets implemented.

**Comment by Mc-Pain** [comment]
There are several reasons why this PR is marked WIP.

> * Obviously, as you've mentioned, performance. Trying to brute-force Rayleigh-Mie is pretty much impossible to do outside of an offline raytracer, which is what that article was originally written for. There are plenty of papers available discussing ways to make it more performant for real-time rendering, but more on that later.

That's one of them.
Looking at the code and trying to figure out what went wrong I became aware of what this model does and how it works. Myself, without looking at the original paper.
The quality of image could be performed via increasing numSamples, but I think it could be easier to solve an integral and get a single solution with given parameters, reducing complexity from O(n^2) to O(n).

> * The reference Rayleigh/Mie coefficients used by most implementations (and this one specifically) are (AFAIK) only appropriate for an oxygenated / earth-like atmosphere and will not produce visually correct results in CO2 / Methane atmospheres or in different atmospheric densities. Without some approximation to generate semi-physically-based Rayleigh / Mie coefficients there's not much point to introducing "realistic" atmospheric scattering.

Atmosphere's color can be used for adjusting these coefficients (does it change between planets with similar composition?)

> Have you verified if "aerial perspective" AKA views from space work correctly? Depending on the implementation of the Nishita model those are sometimes not handled correctly.

That's another one but it might stuck without help of whoever wrote this shader or know what some vars mean and do.

**Comment by Web-eWorks** [comment]
> but I think it could be easier to solve an integral and get a single solution with given parameters, reducing complexity from O(n^2) to O(n).

> That's another one but it might stuck without help of whoever wrote this shader or know what some vars mean and do.

Read this: https://developer.nvidia.com/gpugems/gpugems2/part-ii-shading-lighting-and-shadows/chapter-16-accurate-atmospheric-scattering - it's another implementation of the same atmospheric scattering model using an LUT instead of brute-forcing the integral. You'll have to transcribe the code from Cg if you want to crib off of it, but I recommend implementing it from the article's description instead.

**Comment by Mc-Pain** [comment]
> You'll have to transcribe the code from Cg if you want to crib off of it, but I recommend implementing it from the article's description instead.

Read the whole article. Not understood where's the difference between this and mine (except for the LUP mentioned above)

**Comment by Mc-Pain** [comment]
I am very suspicious of "density function along the vector" being similar to normal distribution.

**Comment by impaktor** [comment]
If we include this, it might even be "too much" to put the on/off in the settings menu, might belong in debug menu instead, for developers / people who want to render a scene for screen shot. just a thought.

**Comment by Mc-Pain** [comment]
Instead, I built my own model for atmospheric density.

Found out this:
1) density from outside atmosphere can be approximated: k*exp(h*b) where h = distance of closest point of the ray to geosphereCenter
2) density from inside could be multiplied by erf(t), where t is parameter from -e to e:
t = -e - no atmosphere along the ray
t = 0 - half of atmosphere along the ray (we're at the closest point)
t = e - everything is along the ray.

Pretty close to each other
pioneer-atmosphere

**Comment by fluffyfreak** [comment]
This is a very exciting development!
There was, long ago, another attempt to do [atmospheric scattering](https://github.com/fluffyfreak/pioneer/tree/scattering) but it was... 😱 12 years ago...

**Comment by Mc-Pain** [comment]
Pushed some optimization but it looks like I messed up some math and scattering looks unexpectedly different.

Could use some fresh look at this

**Comment by impaktor** [comment]
@Mc-Pain Usually, devs don't even have time for their own code, so best to "self review" / leave it to future self.

**Comment by Mc-Pain** [comment]
> @Mc-Pain Usually, devs don't even have time for their own code, so best to "self review" / leave it to future self.

well, I'll try it myself then. Got to my high-end PC, look what happens at very high sample count

**Comment by Web-eWorks** [comment]
I intend to review this PR at some point in the near future, but I've not yet had a chance to get to it.

**Comment by Mc-Pain** [comment]
Looks like it works now as intended...
At close orbits

изображение

**Comment by nozmajner** [comment]
On my rig (Win10, nvidia 3070ti) it seems to run at 60fps at all times.
And I think it looks beautiful. Although on Earth at dusk and dawn the sun is blown out by the glow. Not sure if that's how it supposed to be:

https://github.com/pioneerspacesim/pioneer/assets/4182678/4c780104-bcff-462d-a61f-46d76772de11

**Review by Web-eWorks — CHANGES_REQUESTED** [review]
Overall, I really like what you've done here to implement this model, even if I'm not clear on the details of how you've implemented it mathematically. I still have some reservations about the physical correctness of the model (in terms of incoming / outgoing units and getting "realistic" atmosphere behavior for a variety of atmosphere compositions) and how well it will handle being ported to a high-dynamic-range rendering pipeline, but this PR definitely gets my approval in terms of massively improving the graphical fidelity of Pioneer!

The rest of this review contains a running list of issues I've noticed while testing this PR - while I won't require every one of them to be solved 100% to merge this PR, I think each issue definitely detracts from the visual quality and believability of the effect and would like to see them addressed before merge.

-----

I'm noticing at low / grazing viewing angles the atmosphere well beyond the terminator is illuminated despite the impossibility of light reaching there without first illuminating the higher atmosphere:

image

Sol is well below the horizon here, and the illumination seems to be reasonably correct here:

image

But looking the opposite direction the atmosphere is illuminated where it should otherwise be *well* occluded by the planet.
image

This behavior is not limited to the per-vertex approach and is present in per-pixel mode as well. From further inspection, this appears to be due to the fact that the incident light computation is not handling the effects of the planet self-shadowing and is instead relying on `atmosDiffuse` to darken itself.

Performance seems to be quite good however - the effect in per-pixel mode costs ~1.5ms at 1920x1080 on my 2070 Super and will scale linearly with pixel count.

-----

I am noticing however that while the atmosphere from an aerial perspective looks fairly high-quality, the illumination on terrain and models in the distance is quite jarringly different as the `multi` and `geosphere_terrain` shaders still use the legacy lighting calculations.

Unfortunately, the options there are either to compute incident light and scattering per-pixel for all affected geometry, or to generate per-frame a three-dimensional volume lookup texture computing the optical depth / transmittance term over the view frustum. The first is doable with the existing infrastructure, the second - while much more scalable - will require a potentially major amount of changes to the existing high-level rendering infrastructure to make that LUT available.

-----

It doesn't look like there's any support for varying the Raleigh/Mie `beta` color coefficients yet, which results in a quite unrealistic sky on planets like Mars, which seems to have a much dustier and less blue atmosphere. Even on Earth, there's a very significant green tinge to the edge of the atmosphere at all times of day.

image

Changing the Rayleigh `beta` coefficient to `5.5e-6, 13.0e-6, 22.4e-6` seems to produce a more realistic sky while landed on Earth:

image

I also added a tonemapping step to the end of `rayleigh_accurate.frag` borrowed from https://www.shadertoy.com/view/slSXRW and https://www.shadertoy.com/view/4dBcD1, which gets rid of the "blown-out" sky effect:

```glsl
vec3 jodieReinhardTonemap(vec3 c){
    float l = dot(c, vec3(0.2126, 0.7152, 0.0722));
    vec3 tc=c/(c+1.);
    return mix(c/(l+1.),tc,tc);
}

void main(void)
{
//...

	// 40.0 stands as an intensity scalar to get relatively-correct brightness levels in an LDR pipeline
	specularHighlight *= vec3(40.0);
	// produces a more vibrant / saturated final image
	specularHighlight = pow(specularHighlight, vec3(1.3));

	vec3 final_color = atmosDiffuse.xyz * specularHighlight;
	frag_color = vec4(jodieReinhardTonemap(final_color), 1.0);
}
```

image

There's an alternate version which omits the `pow(specular, vec3(1.3)` and `jodieReinhardTonemap`, instead using `1.0 - exp(-final_color)` as the tonemap for a slightly brighter and more neutral / desaturated image.

```glsl
	specularHighlight *= vec3(40.0);
	vec3 final_color = atmosDiffuse.xyz * specularHighlight;
	frag_color = vec4(1.0 - exp(-final_color), 1.0);
```

image

-----

I'm also noticing the implementation here doesn't handle a light absorption term, which means sunrise and sunset lacks that characteristic "red hue". [This shadertoy](https://www.shadertoy.com/view/wlBXWK) is MIT-licensed and has a fairly clear implementation of atmospheric absorption as well as a good set of coefficients for an earthlike atmosphere if you're interested in adding that to the model.

I've used its Rayleigh coefficients above to produce the more realistic sky color you see in the screenshots.

**Comment by Web-eWorks** [comment]
And how could I forget - part of the color issues are caused by the fact that the blending is taking place in sRGB space instead of linear space!

With an sRGB->Linear conversion of `uLight[i].diffuse` and a matching Linear->sRGB conversion at the end of the pipeline, the intensity scalar can be reduced down to `20.0` and you get much more natural colors - still no red hue shift when the sun is very close to the horizon though.

With the `jodieReinhardTonemap` operator:
image

With the `1.0 - exp(-final_color)` operator and no `pow(specularHighlight, 1.3)`:
image

-----

To better replicate the color grading and look of the sRGB method while blending in linear space, I set the intensity to 20.0, bumped the exponentiation term up to 1.5 from 1.3 and recorded this video (warning: 30MB!):

https://github.com/pioneerspacesim/pioneer/assets/4218491/7a948736-c0fc-48e2-80f0-7a5bf4f51f97

(Music: distant_home_v2_nonlooping.ogg, soon to be added to the Pioneer soundtrack!)

I'd be more than happy to provide a patch or push a commit to this branch implementing these changes if you're interested. With blending happening in linear space, the atmosphere will need some kind of debanding effect but that shouldn't be terribly difficult to implement.

-----

Overall I'm *extremely* happy with the visual effect achieved by this branch now that all of the puzzle pieces are starting to come together. Thank you for persevering with this effort @Mc-Pain!

**Comment by Mc-Pain** [comment]
> I'm noticing at low / grazing viewing angles the atmosphere well beyond the terminator is illuminated despite the impossibility of light reaching there without first illuminating the higher atmosphere:

That definitely looks like being illuminated from behind the planet.
But it is actually illumination from *closer* layers which are still being under direct light.

> But looking the opposite direction the atmosphere is illuminated where it should otherwise be well occluded by the planet.

I suspect the same thing as above. We need to match the colors of atmosphere and terrain to resolve this.

**Comment by Mc-Pain** [comment]
> It doesn't look like there's any support for varying the Raleigh/Mie beta color coefficients yet, which results in a quite unrealistic sky on planets like Mars, which seems to have a much dustier and less blue atmosphere. Even on Earth, there's a very significant green tinge to the edge of the atmosphere at all times of day.

I suggest instead of manipulating `beta` coefficients make them change different colors instead of pure RGB parts.

**Comment by Web-eWorks** [comment]
> I suggest instead of manipulating `beta` coefficients make them change different colors instead of pure RGB parts.

The only way I see for that to make sense is to redefine the wavelength of light that the scattering coefficients are defined in terms of. The `beta` coefficients currently used are defined as the scattering of three extremely specific wavelengths of light (pure red, green, and blue) in specifically an Earth atmosphere (~70 N2, 21% O2, etc...) at sea level.

To represent other planetary atmospheres, we can vary the color of the sunlight and its intensity, but also need to adjust the scattering coefficients of the atmosphere to account for the different composition of gases which have a different effect on the scattering of those specific wavelengths of light.

File: src/galaxy/SystemBody.cpp
Comment (Mc-Pain) [reply src/galaxy/SystemBody.cpp:310]: To handle scattering, we need to do two things:
1) Calculate optical depth along the view ray (from observer to a certain point)
2) Calculate optical depth along the light ray (from point defined above to a light source)

First version of model: take `k=16` uniformly distributed points along the ray, calculate "density density" at each point and multiply by segment length, which involves a nested loop (depth=2)

To get this faster by computing density along the ray in one operation I found out that density function looks pretty like uniform distribution.
By similar digging I found out that overall density along the ray is some exponent.

So the algorithm was:
1) Get total density along the ray:
`total_dens = k*e^(b*h)`
2) Get "density ratio" between given points
`part_dens = total_dens * (erf(t1) - erf(t0))`

File: src/graphics/Drawables.cpp
Comment (Mc-Pain) [reply src/graphics/Drawables.cpp:485]: This was agreed to get rid of.

**Comment by Mc-Pain** [comment]
> The only way I see for that to make sense is to redefine the wavelength of light that the scattering coefficients are defined in terms of. The `beta` coefficients currently used are defined as the scattering of three extremely specific wavelengths of light (pure red, green, and blue) in specifically an Earth atmosphere (~70 N2, 21% O2, etc...) at sea level.

Colors could be defined from here:
https://www.shadertoy.com/view/NlGXzz

**Comment by Web-eWorks** [comment]
Other than the aliasing issues on the edge of the terrain (I assume this was taken with AA off...) that first screenshot looks absolutely beautiful.

**Comment by fluffyfreak** [comment]
The planet ring shaders look very self contained, you could maybe do those as a separate PR just to keep this one a little smaller if you wanted too.

**Comment by Mc-Pain** [comment]
> The planet ring shaders look very self contained, you could maybe do those as a separate PR just to keep this one a little smaller if you wanted too.

I'll cherry-pick this then

File: data/shaders/opengl/basesphere_uniforms.glsl
Comment (Web-eWorks) [inline data/shaders/opengl/basesphere_uniforms.glsl:29]: These should be described with comments indicating their purpose and their values.

File: data/shaders/opengl/rayleigh.glsl
Comment (Web-eWorks) [inline data/shaders/opengl/rayleigh.glsl:30]: This should be described in more detail, at minimum something along the lines of "error function to approximate (in/out)-scattering integral over the ray between the sample position and the light source".

File: data/shaders/opengl/rayleigh.glsl
Comment (Web-eWorks) [inline data/shaders/opengl/rayleigh.glsl:96]: I'd recommend lowering the number of samples used by the skip ray as much as possible without compromising the quality of the resulting scattering. In my testing a high sample count here did not significantly contribute to the visual quality of atmospheres.

File: data/shaders/opengl/rayleigh.glsl
Comment (Web-eWorks) [inline data/shaders/opengl/rayleigh.glsl:233]: If this uses the same Mie phase function as the planet ring PR, please convert this to a call to that function once it is merged.

File: src/galaxy/SystemBody.cpp
Comment (Web-eWorks) [inline src/galaxy/SystemBody.cpp:266]: I'd strongly prefer this function be annotated with a description of both the math inside as well as its purpose.

File: src/galaxy/SystemBody.cpp
Comment (Web-eWorks) [inline src/galaxy/SystemBody.cpp:287]: I'd strongly prefer this function be annotated with a description of both the math inside as well as its purpose.

**Review by Web-eWorks — CHANGES_REQUESTED** [review]
First of all, thank you for taking the time and effort to implement this as far as you have!

Regarding prior discussion on IRC about this PR's state, I'm generally willing to merge this for the upcoming release in its current condition as an experimental option. While it still requires significant further improvements to match terrain rendering with atmosphere rendering before I consider it a feature-complete replacement for our current atmosphere rendering, those improvements can be added in after its inclusion in the upcoming release.

To that end, I've pushed several commits to add "hot swapping" the used atmosphere shader from the settings menu and to clearly indicate that the Rayleigh/Mie shader is experimental.

I'd like to see the attached comments be addressed before merging, but if you consider this in a good enough state to merge for the upcoming release then I'll sign off on it.

**Comment by impaktor** [comment]
Just checking: Could this PR cause crashes? My main thinking is: if someone opens a bug report, and attach output.txt, will we see they've enabled the "Experimental" option?

File: data/shaders/opengl/rayleigh.glsl
Comment (Mc-Pain) [reply data/shaders/opengl/rayleigh.glsl:96]: I'll leave 8

File: data/shaders/opengl/basesphere_uniforms.glsl
Comment (Web-eWorks) [inline data/shaders/opengl/basesphere_uniforms.glsl:29]: Note: in all literature I'm aware of, the distance over which a quantity reduces by a factor of _e_ is called [Scale Height](https://en.wikipedia.org/wiki/Scale_height). If this is in fact the atmospheric scale height for Rayleigh/Mie scattering, I'd recommend this parameter be renamed such in all cases as `baricStep` is unclear.

**Comment by Web-eWorks** [comment]
> Just checking: Could this PR cause crashes? My main thinking is: if someone opens a bug report, and attach output.txt, will we see they've enabled the "Experimental" option?

Very unlikely. This PR mainly makes changes to shaders and the experimental option simply influences which shader is chosen.

**Review by Web-eWorks — APPROVED** [review]
@Mc-Pain please go ahead and start squashing related / iterative-work commits down into feature-level commits. Ideally all "rename" and "fix" commits related to the new rayleigh scattering code should be condensed into a small number of commits adding a major feature. Generally unrelated or standalone commits (e.g. "Add experimental label to Rayleigh/Mie scattering option") can be left as-is.

I generally consider this PR as approved, but the commit history will need to be cleaned up before merge. The sooner the better - once the history has been cleaned up I consider this PR good to merge for whoever is able to hit the button first.

**Comment by Mc-Pain** [comment]
image
image
image

**Comment by fluffyfreak** [comment]
Hi @Mc-Pain could you open a new bug report with this image in, where it happened and some info about your setup please?

**Comment by sturnclaw** [comment]
I believe these are work-in-progress screenshots of a future improvement to the shader, not a bug with the current shader.

**Comment by Mc-Pain** [comment]
> I believe these are work-in-progress screenshots of a future improvement to the shader, not a bug with the current shader.

Yes, I had nowhere else to upload those

**Pull request merged**
